File: packages/hap-csharp/src/Verify.cs
using System.Text;
using System.Text.Json;
using NSec.Cryptography;

namespace BlueScroll.Hap;

/// <summary>
/// HAP claim verification functions.
/// </summary>
public class HapVerifier : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly bool _ownsHttpClient;

    /// <summary>
    /// Creates a new HAP verifier.
    /// </summary>
    /// <param name="httpClient">Optional HTTP client to use</param>
    public HapVerifier(HttpClient? httpClient = null)
    {
        if (httpClient != null)
        {
            _httpClient = httpClient;
            _ownsHttpClient = false;
        }
        else
        {
            _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
            _ownsHttpClient = true;
        }
    }

    /// <summary>
    /// Fetches the public keys from a VA's well-known endpoint.
    /// </summary>
    /// <param name="issuerDomain">The VA's domain (e.g., "ballista.io")</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>The VA's public key configuration</returns>
    public async Task<HapWellKnown> FetchPublicKeysAsync(string issuerDomain, CancellationToken cancellationToken = default)
    {
        var url = $"https://{issuerDomain}/.well-known/hap.json";
        var response = await _httpClient.GetStringAsync(url, cancellationToken);
        return JsonSerializer.Deserialize<HapWellKnown>(response) ?? new HapWellKnown();
    }

    /// <summary>
    /// Fetches and verifies a HAP claim from a VA.
    /// </summary>
    /// <param name="hapId">The HAP ID to verify</param>
    /// <param name="issuerDomain">The VA's domain</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>The verification response from the VA</returns>
    public async Task<VerificationResponse> FetchClaimAsync(string hapId, string issuerDomain, CancellationToken cancellationToken = default)
    {
        if (!Hap.IsValidHapId(hapId))
        {
            return new VerificationResponse { Valid = false, Error = "invalid_format" };
        }

        var url = $"https://{issuerDomain}/api/v1/verify/{hapId}";
        var response = await _httpClient.GetStringAsync(url, cancellationToken);
        return JsonSerializer.Deserialize<VerificationResponse>(response) ?? new VerificationResponse();
    }

    /// <summary>
    /// Verifies a JWS signature against a VA's public keys.
    /// </summary>
    /// <param name="jws">The JWS compact serialization string</param>
    /// <param name="issuerDomain">The VA's domain to fetch public keys from</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>Verification result with decoded claim if valid</returns>
    public async Task<SignatureVerificationResult> VerifySignatureAsync(string jws, string issuerDomain, CancellationToken cancellationToken = default)
    {
        try
        {
            var wellKnown = await FetchPublicKeysAsync(issuerDomain, cancellationToken);

            var parts = jws.Split('.');
            if (parts.Length != 3)
            {
                return new SignatureVerificationResult { Valid = false, Error = "Invalid JWS format" };
            }

            var headerJson = Base64UrlDecode(parts[0]);
            var header = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(headerJson);
            var kid = header?.GetValueOrDefault("kid").GetString();

            if (string.IsNullOrEmpty(kid))
            {
                return new SignatureVerificationResult { Valid = false, Error = "JWS header missing kid" };
            }

            var jwk = wellKnown.Keys.FirstOrDefault(k => k.Kid == kid);
            if (jwk == null)
            {
                return new SignatureVerificationResult { Valid = false, Error = $"Key not found: {kid}" };
            }

            // Decode the public key
            var publicKeyBytes = Base64UrlDecodeBytes(jwk.X);
            var algorithm = SignatureAlgorithm.Ed25519;
            var publicKey = PublicKey.Import(algorithm, publicKeyBytes, KeyBlobFormat.RawPublicKey);

            // Verify the signature
            var signingInput = Encoding.ASCII.GetBytes($"{parts[0]}.{parts[1]}");
            var signature = Base64UrlDecodeBytes(parts[2]);

            if (!algorithm.Verify(publicKey, signingInput, signature))
            {
                return new SignatureVerificationResult { Valid = false, Error = "Signature verification failed" };
            }

            // Decode the payload
            var payloadJson = Base64UrlDecode(parts[1]);
            var claim = JsonSerializer.Deserialize<HumanEffortClaim>(payloadJson);

            if (claim?.Iss != issuerDomain)
            {
                return new SignatureVerificationResult
                {
                    Valid = false,
                    Error = $"Issuer mismatch: expected {issuerDomain}, got {claim?.Iss}"
                };
            }

            return new SignatureVerificationResult { Valid = true, Claim = claim };
        }
        catch (Exception ex)
        {
            return new SignatureVerificationResult { Valid = false, Error = ex.Message };
        }
    }

    /// <summary>
    /// Fully verifies a HAP claim: fetches from VA and optionally verifies signature.
    /// </summary>
    /// <param name="hapId">The HAP ID to verify</param>
    /// <param name="issuerDomain">The VA's domain</param>
    /// <param name="verifySignature">Whether to verify the cryptographic signature</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>The claim if valid, null if not found or invalid</returns>
    public async Task<HumanEffortClaim?> VerifyHapClaimAsync(
        string hapId,
        string issuerDomain,
        bool verifySignature = true,
        CancellationToken cancellationToken = default)
    {
        var response = await FetchClaimAsync(hapId, issuerDomain, cancellationToken);

        if (!response.Valid)
        {
            return null;
        }

        if (verifySignature && !string.IsNullOrEmpty(response.Jws))
        {
            var sigResult = await VerifySignatureAsync(response.Jws, issuerDomain, cancellationToken);
            if (!sigResult.Valid)
            {
                return null;
            }
        }

        return response.Claims;
    }

    /// <summary>
    /// Extracts the HAP ID from a verification URL.
    /// </summary>
    /// <param name="url">The verification URL</param>
    /// <returns>The HAP ID or null if not found</returns>
    public static string? ExtractHapIdFromUrl(string url)
    {
        try
        {
            var uri = new Uri(url);
            var parts = uri.AbsolutePath.Split('/');
            var lastPart = parts.LastOrDefault();

            return Hap.IsValidHapId(lastPart ?? "") ? lastPart : null;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Checks if a claim is expired.
    /// </summary>
    /// <param name="claim">The HAP claim to check</param>
    /// <returns>true if the claim has an exp field and is expired</returns>
    public static bool IsClaimExpired(HapClaim claim)
    {
        if (string.IsNullOrEmpty(claim.Exp))
        {
            return false;
        }

        try
        {
            var expTime = DateTime.Parse(claim.Exp).ToUniversalTime();
            return expTime < DateTime.UtcNow;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Checks if the claim target matches the expected company.
    /// </summary>
    /// <param name="claim">The HAP claim to check</param>
    /// <param name="companyDomain">The expected company domain</param>
    /// <returns>true if the claim's target domain matches</returns>
    public static bool IsClaimForCompany(HumanEffortClaim claim, string companyDomain)
    {
        return claim.To?.Domain == companyDomain;
    }

    /// <summary>
    /// Checks if an employer commitment claim matches the expected company.
    /// </summary>
    /// <param name="claim">The employer commitment claim to check</param>
    /// <param name="companyDomain">The expected company domain</param>
    /// <returns>true if the claim's employer domain matches</returns>
    public static bool IsClaimForCompany(EmployerCommitmentClaim claim, string companyDomain)
    {
        return claim.Employer?.Domain == companyDomain;
    }

    private static string Base64UrlDecode(string input)
    {
        return Encoding.UTF8.GetString(Base64UrlDecodeBytes(input));
    }

    private static byte[] Base64UrlDecodeBytes(string input)
    {
        var output = input.Replace('-', '+').Replace('_', '/');
        switch (output.Length % 4)
        {
            case 2: output += "=="; break;
            case 3: output += "="; break;
        }
        return Convert.FromBase64String(output);
    }

    public void Dispose()
    {
        if (_ownsHttpClient)
        {
            _httpClient.Dispose();
        }
    }
}

File: packages/go/verify.go
package humanattestation

import (
	"context"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/go-jose/go-jose/v4"
)

// DefaultTimeout is the default HTTP request timeout
const DefaultTimeout = 10 * time.Second

// VerifyOptions configures verification behavior
type VerifyOptions struct {
	// HTTPClient allows using a custom HTTP client
	HTTPClient *http.Client
	// Timeout for HTTP requests (default: 10s)
	Timeout time.Duration
	// VerifySignature controls whether to verify the cryptographic signature
	VerifySignature bool
}

// DefaultVerifyOptions returns options with sensible defaults
func DefaultVerifyOptions() VerifyOptions {
	return VerifyOptions{
		HTTPClient:      http.DefaultClient,
		Timeout:         DefaultTimeout,
		VerifySignature: true,
	}
}

// IsValidID validates a HAP ID format
func IsValidID(id string) bool {
	return IDRegex.MatchString(id)
}

// FetchPublicKeys fetches the public keys from a VA's well-known endpoint
func FetchPublicKeys(ctx context.Context, issuerDomain string, opts VerifyOptions) (*WellKnown, error) {
	if opts.HTTPClient == nil {
		opts.HTTPClient = http.DefaultClient
	}
	if opts.Timeout == 0 {
		opts.Timeout = DefaultTimeout
	}

	ctx, cancel := context.WithTimeout(ctx, opts.Timeout)
	defer cancel()

	url := fmt.Sprintf("https://%s/.well-known/hap.json", issuerDomain)
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Accept", "application/json")

	resp, err := opts.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch public keys: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch public keys: HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var wellKnown WellKnown
	if err := json.Unmarshal(body, &wellKnown); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &wellKnown, nil
}

// FetchClaim fetches and verifies a HAP claim from a VA
func FetchClaim(ctx context.Context, hapID, issuerDomain string, opts VerifyOptions) (*VerificationResponse, error) {
	if !IsValidID(hapID) {
		return &VerificationResponse{Valid: false, Error: "invalid_format"}, nil
	}

	if opts.HTTPClient == nil {
		opts.HTTPClient = http.DefaultClient
	}
	if opts.Timeout == 0 {
		opts.Timeout = DefaultTimeout
	}

	ctx, cancel := context.WithTimeout(ctx, opts.Timeout)
	defer cancel()

	url := fmt.Sprintf("https://%s/api/v1/verify/%s", issuerDomain, hapID)
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Accept", "application/json")

	resp, err := opts.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch claim: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var verifyResp VerificationResponse
	if err := json.Unmarshal(body, &verifyResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &verifyResp, nil
}

// VerifySignature verifies a JWS signature against a VA's public keys
func VerifySignature(ctx context.Context, jwsString, issuerDomain string, opts VerifyOptions) (*SignatureVerificationResult, error) {
	// Fetch public keys
	wellKnown, err := FetchPublicKeys(ctx, issuerDomain, opts)
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: err.Error()}, nil
	}

	// Parse the JWS
	jws, err := jose.ParseSigned(jwsString, []jose.SignatureAlgorithm{jose.EdDSA})
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("failed to parse JWS: %v", err)}, nil
	}

	// Get the key ID from the header
	if len(jws.Signatures) == 0 {
		return &SignatureVerificationResult{Valid: false, Error: "no signatures in JWS"}, nil
	}
	kid := jws.Signatures[0].Header.KeyID
	if kid == "" {
		return &SignatureVerificationResult{Valid: false, Error: "JWS header missing kid"}, nil
	}

	// Find the matching key
	var jwk *JWK
	for _, k := range wellKnown.Keys {
		if k.Kid == kid {
			jwk = &k
			break
		}
	}
	if jwk == nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("key not found: %s", kid)}, nil
	}

	// Decode the public key
	xBytes, err := base64.RawURLEncoding.DecodeString(jwk.X)
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("failed to decode public key: %v", err)}, nil
	}
	publicKey := ed25519.PublicKey(xBytes)

	// Verify the signature
	payload, err := jws.Verify(publicKey)
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("signature verification failed: %v", err)}, nil
	}

	// Parse the payload
	var claim Claim
	if err := json.Unmarshal(payload, &claim); err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("failed to parse claim: %v", err)}, nil
	}

	// Verify issuer matches
	if claim.Iss != issuerDomain {
		return &SignatureVerificationResult{
			Valid: false,
			Error: fmt.Sprintf("issuer mismatch: expected %s, got %s", issuerDomain, claim.Iss),
		}, nil
	}

	return &SignatureVerificationResult{Valid: true, Claim: &claim}, nil
}

// VerifyClaim fully verifies a HAP claim: fetches from VA and optionally verifies signature
func VerifyClaim(ctx context.Context, hapID, issuerDomain string, opts ...VerifyOptions) (*Claim, error) {
	var opt VerifyOptions
	if len(opts) > 0 {
		opt = opts[0]
	} else {
		opt = DefaultVerifyOptions()
	}

	// Fetch the claim
	resp, err := FetchClaim(ctx, hapID, issuerDomain, opt)
	if err != nil {
		return nil, err
	}

	// Check if valid
	if !resp.Valid {
		return nil, nil
	}

	// Optionally verify the signature
	if opt.VerifySignature && resp.JWS != "" {
		sigResult, err := VerifySignature(ctx, resp.JWS, issuerDomain, opt)
		if err != nil {
			return nil, err
		}
		if !sigResult.Valid {
			return nil, nil
		}
	}

	return resp.Claim, nil
}

// ExtractIDFromURL extracts the HAP ID from a verification URL
func ExtractIDFromURL(urlStr string) string {
	parsed, err := url.Parse(urlStr)
	if err != nil {
		return ""
	}

	parts := strings.Split(parsed.Path, "/")
	if len(parts) == 0 {
		return ""
	}

	lastPart := parts[len(parts)-1]
	if IsValidID(lastPart) {
		return lastPart
	}

	return ""
}

// IsClaimExpired checks if a claim is expired
func IsClaimExpired(claim *Claim) bool {
	if claim.Exp == "" {
		return false
	}

	expTime, err := time.Parse(time.RFC3339, claim.Exp)
	if err != nil {
		return false
	}

	return expTime.Before(time.Now())
}

// IsClaimForRecipient checks if the claim target matches the expected recipient
func IsClaimForRecipient(claim *Claim, recipientDomain string) bool {
	return claim.To.Domain == recipientDomain
}
